src/device_manager.rs

```diff
@@ -0,0 +1,431 @@
+        if self.devices.get(&(dev.instance_id)).is_some() {
+            return Err(Error::Exist);
+        }
+        self.devices.insert(dev.instance_id, dev);
+        Ok(())
+    }
+
+    fn remove(&mut self, instance_id: u32) -> Option<DeviceDescriptor> {
+        self.devices.remove(&instance_id)
+    }
+
+    fn device_descriptor(
+        &self,
+        id: u32,
+        name: String,
+        dev: Arc<Mutex<dyn Device>>,
+        parent_bus: Option<Arc<Mutex<dyn Device>>>,
+        resources: Vec<IoResource>,
+        irq: Option<IrqResource>,
+    ) -> DeviceDescriptor {
+        DeviceDescriptor::new(id, name, dev.clone(), parent_bus, resources, irq)
+    }
+
+    // Allocate IO and instance id resources.
+    // Return a Result wrapper of instance id.
+    fn allocate_resources(&mut self, resources: &mut Vec<IoResource>) -> Result<(u32)> {
+        let id = self
+            .resource
+            .lock()
+            .expect("failed to acquire lock.")
+            .allocate_instance_id()
+            .map_err(Error::InstanceIdAllocate)?;
+
+        for res in resources.iter_mut() {
+            match res.res_type {
+                IoType::Pio => {
+                    if res.addr.is_none() {
+                        return Err(Error::NonePIORequest);
+                    }
+                    res.addr = Some(
+                        self.resource
+                            .lock()
+                            .expect("failed to acquire lock")
+                            .allocate_io_addresses(res.addr.unwrap(), res.size)
+                            .map_err(Error::AddressAllocate)?,
+                    );
+                }
+                IoType::PhysicalMmio | IoType::Mmio => {
+                    res.addr = Some(
+                        self.resource
+                            .lock()
+                            .expect("failed to acquire lock")
+                            .allocate_mmio_addresses(res.addr, res.size)
+                            .map_err(Error::AddressAllocate)?,
+                    )
+                }
+            }
+        }
+        Ok(id)
+    }
+
+    fn free_resources(&mut self, resources: &[IoResource], id: u32) {
+        for res in resources.iter() {
+            match res.res_type {
+                IoType::Pio => self
+                    .resource
+                    .lock()
+                    .expect("failed to acquire lock")
+                    .free_io_addresses(res.addr.unwrap(), res.size),
+                IoType::PhysicalMmio | IoType::Mmio => self
+                    .resource
+                    .lock()
+                    .expect("failed to acquire lock")
+                    .free_mmio_addresses(res.addr.unwrap(), res.size),
+            }
+        }
+        self.resource
+            .lock()
+            .expect("failed to acquire lock")
+            .free_instance_id(id);
+    }
+
+    fn register_resources(
```

Comment:
unregister_resoures() may be needed too.

Reply:
So let's create a explicit function for the resources freeing, instead of inside unregister_device().